dcpy/utils/validation.py

```diff
@@ -0,0 +1,97 @@
+import inspect
+from types import UnionType
+from typing import Callable, get_args, get_origin, Literal
+
+
+def _isempty(obj) -> bool:
+    return obj == inspect._empty
+
+
+def _isinstance(obj, cls):
+    """
+    _isinstance extends behavior of basic python `isinstance` to allow for
+    parameterized generics (i.e. check if an object is dict[str, str])
+    """
+    if get_origin(cls) == Literal:
+        return obj in get_args(cls)
+    elif get_origin(cls) == UnionType:
+        return any([_isinstance(obj, param) for param in get_args(cls)])
+    elif get_origin(cls) == list:
+        param = get_args(cls)[0]
+        return isinstance(obj, list) and all([_isinstance(l, param) for l in obj])
+    elif get_origin(cls) == dict:
+        keytype = get_args(cls)[0]
+        valtype = get_args(cls)[1]
+        return (
+            isinstance(obj, dict)
+            and all([_isinstance(key, keytype) for key in obj])
+            and all([_isinstance(obj[key], valtype) for key in obj])
+        )
+    else:
+        return isinstance(obj, cls)
+
+
+# This maybe belongs in a more pure util section of code
+def validate_function_args(
+    function: Callable,
+    kwargs: dict,
+    raise_error=False,
+    ignore_args: list[str] | None = None,
+) -> dict[str, str]:
+    """
+    Given a function and dict containing kwargs, validates that kwargs satiffy the
+    signature of the function. Violations are returned as a dict, with argument names
+    as keys and types of violation as value. Types of violation can be either a missing argument,
+    an argument of the wrong type, or an unexpected argument.
+    If raise_error flag supplied, raises error instead of returning dict of violations.
+    """
+    ignore_args = ignore_args or []
+    sig = inspect.signature(function)
+
+    pos = [
+        a
+        for a in sig.parameters.values()
+        if a.kind == a.VAR_POSITIONAL or a.kind == a.POSITIONAL_ONLY
+    ]
+    if pos:
+        raise TypeError(
```

Comment:
Is this inconsistent with the `raise_error` arg? In the spirit of "a function should do one thing only" maybe it's best to not raise errors and just return the violations?

Reply:
Maybe, though this is a different error - the function itself being validated is not a format that allows validation with this function, rather than some sort of error with the arguments (which is what is supposed to be validated here)

Reply:
Gotcha - I do think that's somewhat unexpected behavior for a function called `validate_function_args`. My gut is that supplying args vs kwargs here isn't so different as to warrant such a change in behavior (ie, raising error vs return errors) but I get the distinction

Reply:
Yeah mainly I was thinking of my use case - I can play around and see if I can just make it slightly more flexible, because I agree that on the face of it, it doesn't necessarily warrant different behavior